src/rsa/public/modulus.rs

```diff
@@ -0,0 +1,69 @@
+use crate::{arithmetic::bigint, bits, error, rsa::N};
+use core::ops::RangeInclusive;
+
+pub struct Modulus {
+    value: bigint::Modulus<N>,
+    bits: bits::BitLength,
+}
+
+impl core::fmt::Debug for Modulus {
+    fn fmt(&self, fmt: &mut ::core::fmt::Formatter) -> Result<(), ::core::fmt::Error> {
+        self.value.fmt(fmt)
+    }
+}
+
+impl Modulus {
+    pub(in crate::rsa) fn from_be_bytes(
+        n: untrusted::Input,
+        allowed_bit_lengths: RangeInclusive<bits::BitLength>,
+    ) -> Result<Self, error::KeyRejected> {
+        // This is an incomplete implementation of NIST SP800-56Br1 Section
+        // 6.4.2.2, "Partial Public-Key Validation for RSA." That spec defers
+        // to NIST SP800-89 Section 5.3.3, "(Explicit) Partial Public Key
+        // Validation for RSA," "with the caveat that the length of the modulus
+        // shall be a length that is specified in this Recommendation." In
+        // SP800-89, two different sets of steps are given, one set numbered,
+        // and one set lettered. TODO: Document this in the end-user
+        // documentation for RSA keys.
+
+        let min_bits = *allowed_bit_lengths.start();
+        let max_bits = *allowed_bit_lengths.end();
+
+        // `pkcs1_encode` depends on this not being small. Otherwise,
+        // `pkcs1_encode` would generate padding that is invalid (too few 0xFF
+        // bytes) for very small keys.
+        const MIN_BITS: bits::BitLength = bits::BitLength::from_usize_bits(1024);
+
+        // Step 3 / Step c for `n` (out of order).
+        let (value, bits) = bigint::Modulus::from_be_bytes_with_bit_length(n)?;
+
+        // Step 1 / Step a. XXX: SP800-56Br1 and SP800-89 require the length of
+        // the public modulus to be exactly 2048 or 3072 bits, but we are more
+        // flexible to be compatible with other commonly-used crypto libraries.
+        assert!(min_bits >= MIN_BITS);
+        let bits_rounded_up =
+            bits::BitLength::from_usize_bytes(bits.as_usize_bytes_rounded_up()).unwrap(); // TODO: safe?
+        if bits_rounded_up < min_bits {
+            return Err(error::KeyRejected::too_small());
+        }
+        if bits > max_bits {
+            return Err(error::KeyRejected::too_large());
+        }
+
+        // TODO: what about this?
```

Comment:
Perhaps this could be checked inside `Key::from_modulus_and_exponent`?

Reply:
I removed the comment here since it is still in `Key`. I do think I need to follow up and clarify (probably remove) that comment after I investigate the history of it.

Reply:
That sounds fine since its technically out of scope for this MR either way.